cpp/oneapi/dal/backend/primitives/stat/test/cov_dpc.cpp

```diff
@@ -0,0 +1,129 @@
+            te::linalg::matrix<Float>::wrap(vars.get_data(), { vars.get_count(), 1 });
+
+        const double eps = std::abs(expected_var) * te::get_tolerance_for_sum<Float>(row_count);
+
+        te::linalg::enumerate_linear(vars_mat, [&](std::int64_t i, Float var) {
+            if (std::abs(double(var) - expected_var) > eps) {
+                CAPTURE(i, var, expected_var);
+                FAIL("Unexpected variance");
+            }
+        });
+    }
+
+    void check_constant_mean(const ndarray<Float, 1>& means,
+                             std::int64_t row_count,
+                             double expected_mean) {
+        const auto means_mat =
+            te::linalg::matrix<Float>::wrap(means.get_data(), { means.get_count(), 1 });
+
+        const double eps = te::get_tolerance<Float>(1e-6, 1e-12);
+
+        te::linalg::enumerate_linear(means_mat, [&](std::int64_t i, Float mean) {
+            if (std::abs((double(mean) - expected_mean) / expected_mean) >= eps) {
+                CAPTURE(i, mean, expected_mean);
+                FAIL("Unexpected mean");
+            }
+        });
+    }
+};
+
+TEMPLATE_TEST_M(cov_test, "correlation on uncorrelated data", "[cor]", float, double) {
+    // DPC++ GEMM used underneath correlation is not supported on GPU
+    SKIP_IF(this->get_policy().is_cpu());
+
+    using float_t = TestType;
+    auto& queue = this->get_queue();
+
+    const float_t diag_element = 10.5;
+
+    const auto df =
+        GENERATE_DATAFRAME(te::dataframe_builder{ 1000000, 100 }.fill_diag(diag_element));
+
+    const auto column_count = df.get_column_count();
+    auto corr = ndarray<float_t, 2>::empty(queue, { column_count, column_count });
+    auto means = ndarray<float_t, 1>::empty(queue, { column_count });
+    auto vars = ndarray<float_t, 1>::empty(queue, { column_count });
+    auto tmp = ndarray<float_t, 1>::empty(queue, { column_count });
+
+    auto [sums, sums_event] = ndarray<float_t, 1>::full(queue, { column_count }, diag_element);
+    const auto data = df.get_table(this->get_policy(), te::table_id::homogen<float_t>());
+
+    correlation(queue, data, sums, corr, means, vars, tmp, { sums_event }).wait_and_throw();
+
+    SECTION("correlation matrix is ones") {
+        this->check_correlation_for_uncorrelated_data(corr);
+    }
+
+    // The upper part of data matrix is diagonal. In diagonal matrix each column contains only one
+    // non-zero element (`diag_element`), so mean and variances for each feature can be computed
+    // trivially using `diag_element` value.
+
+    SECTION("mean is expected") {
+        const double n = df.get_row_count();
+        const double expected_mean = double(diag_element) / n;
+        this->check_constant_mean(means, n, expected_mean);
+    }
+
+    SECTION("variance is expected") {
+        const double n = df.get_row_count();
+        const double d = double(diag_element) * double(diag_element);
+        const double expected_var = (d - d / n) / (n - 1.0);
```

Comment:
Single row case isn't expected, right?

Reply:
Data is generated to have several rows, make sense to add assertion here and separate test case where `n == 1`.